File: Algorithm/prototype_C#/LowLevel.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace RobotAlgorithm
{
    public class LowLevelMap
    {
        private bool[,] mGraph;

        public LowLevelMap(bool[,] graph)
        {
            mGraph = graph;
        }

        public LowLevelMap(char way, char notWay, string graph)
        {
            int width = -1, height = 0;
            List<List<bool>> tempGraph = new List<List<bool>>();

            using (StringReader sr = new StringReader(graph))
            {
                string line;
                int colPos;

                while ((line = sr.ReadLine()) != null)
                {
                    colPos = 0;
                    tempGraph.Add(new List<bool>());

                    foreach (char ch in line)
                    {
                        if (ch == way)
                            tempGraph[height].Add(true);
                        else if (ch == notWay)
                            tempGraph[height].Add(false);
                        else
                            throw new ArgumentException();

                        ++colPos;
                    }

                    if (width == -1)
                        width = colPos;
                    else if (width != colPos)
                        throw new ArgumentException();

                    ++height;
                }
            }

            mGraph = new bool[height, width];
            for (int row = 0; row < height; ++row)
                for (int col = 0; col < width; ++col)
                    mGraph[row, col] = tempGraph[row][col];
        }

        public int Width => mGraph.GetLength(1);
        public int Height => mGraph.GetLength(0);

        public bool this[int row, int col]
        {
            get
            {
                return mGraph[row, col];
            }
        }

        public bool Available(int row, int col)
        {
            if (row < 0 || row >= Height)
                return false;
            if (col < 0 || col >= Width)
                return false;
            return true;
        }
    }

    public class LowLevelRobot
    {
        public const double ERROR_RATE = 0.05;
        private static Random Rand = new Random();

        int mRealX, mRealY;
        LowLevelMap mMap;

        public double X { get; set; }

        public double Y { get; set; }

        public Direction Dir { get; private set; }

        public LowLevelRobot(LowLevelMap map, int x, int y)
        {
            mMap = map;
            X = mRealX = x;
            Y = mRealY = y;
        }

        public void TurnLeft()
        {
            Dir = Dir.TurnLeft();
        }

        public void TurnRight()
        {
            Dir = Dir.TurnRight();
        }

        public bool Move(int l)
        {
            int dx = 0, dy = 0;

            switch (this.Dir)
            {
                case Direction.North:
                    dy = -l;
                    break;
                case Direction.South:
                    dy = l;
                    break;
                case Direction.East:
                    dx = l;
                    break;
                case Direction.West:
                    dx = -l;
                    break;
                default:
                    return false;
            }

            if (!mMap.Available(mRealY + dy, mRealX + dx))
                return false;
            else if (!mMap[mRealY + dy, mRealX + dx])
                return false;

            double fakeDx = Rand.GetError(ERROR_RATE, dx) + dx;
            double fakeDy = Rand.GetError(ERROR_RATE, dy) + dy;

            Console.Write($"Robot Moved: ({mRealX}, {mRealY}) ");
            mRealX += dx;
            mRealY += dy;
            Console.WriteLine($"-> ({mRealX}, {mRealY})");
            X += fakeDx;
            Y += fakeDy;

            return true;
        }

        public Junction CheckNode()
        {
            Junction llj = new Junction();

            if (mMap[mRealY - 1, mRealX])
                llj.North = true;
            if (mMap[mRealY + 1, mRealX])
                llj.South = true;
            if (mMap[mRealY, mRealX + 1])
                llj.East = true;
            if (mMap[mRealY, mRealX - 1])
                llj.West = true;

            return llj;
        }

        public bool MoveToNode(out Junction junction)
        {
            Junction currJunc = CheckNode();

            Console.WriteLine($"Fake X Changed: ({X}, {Y})");
            do
            {
                if (!Move(1))
                    break;
            } while ((currJunc = CheckNode()).Type == NodeType.Straight);
            Console.WriteLine($"-> ({X}, {Y})");

            junction = currJunc;
            return true;
        }
    }
}

File: Algorithm/ported_C++/Sources/LowLevel.cpp
#include "LowLevel.h"

Point RobotPos;
Direction RobotDir = Direction::North;

//-------------------------------------------------------

bool graph[100][100];
int width, height;
const double VIR_ERROR_RATE = 0.05;

int realX, realY;

void InitForDebug(int x, int y)
{
	RobotPos.X = x;
	RobotPos.Y = y;
	realX = x;
	realY = y;
}

bool Available(int row, int col)
{
	if (row < 0 || row >= height)
		return false;
	if (col < 0 || col >= width)
		return false;
	return true;
}

double LowLevel::GetX()
{
	return RobotPos.X;
}

double LowLevel::GetY()
{
	return RobotPos.Y;
}

Point LowLevel::GetPos()
{
	return RobotPos;
}

Direction LowLevel::GetDir()
{
	return RobotDir;
}

void LowLevel::TurnLeft_Raw()
{
	RobotDir = TurnLeft(RobotDir);
}

void LowLevel::TurnRight_Raw()
{
	RobotDir = TurnRight(RobotDir);
}

bool LowLevel::Move_Raw(double l)
{
	int dx = 0, dy = 0;

	switch (RobotDir)
	{
	case Direction::North:
		dy = -(int)l;
		break;
	case Direction::South:
		dy = (int)l;
		break;
	case Direction::East:
		dx = (int)l;
		break;
	case Direction::West:
		dx = -(int)l;
		break;
	default:
		return false;
	}

	if (!Available(realY + dy, realX + dx))
		return false;
	else if (!graph[realY + dy, realX + dx])
		return false;

	double fakeDx = GetError(VIR_ERROR_RATE, dx) + dx;
	double fakeDy = GetError(VIR_ERROR_RATE, dy) + dy;

	realX += dx;
	realY += dy;
	RobotPos.X += fakeDx;
	RobotPos.Y += fakeDy;

	return true;
}

Junction LowLevel::CheckNode_Raw()
{
	Junction llj;

	if (graph[realY - 1][realX])
		llj.North = true;
	if (graph[realY + 1][realX])
		llj.South = true;
	if (graph[realY][realX + 1])
		llj.East = true;
	if (graph[realY][realX - 1])
		llj.West = true;

	return llj;
}

std::pair<bool, Junction> LowLevel::MoveToNode_Raw()
{
	Junction currJunc = CheckNode_Raw();

	do
	{
		if (!Move_Raw(1))
			break;
	} while ((currJunc = CheckNode_Raw()).GetType() == NodeType::Straight);

	return { true, currJunc };
}

//-------------------------------------------------------

std::future<void> LowLevel::TurnLeft()
{
	return std::async(std::launch::async, TurnLeft_Raw);
}

std::future<void> LowLevel::TurnRight()
{
	return std::async(std::launch::async, TurnRight_Raw);
}

std::future<bool> LowLevel::Move(double l)
{
	return std::async(std::launch::async, Move_Raw, l);
}

std::future<Junction> LowLevel::CheckNode()
{
	return std::async(std::launch::async, CheckNode_Raw);
}

std::future<std::pair<bool, Junction>> LowLevel::MoveToNode()
{
	return std::async(std::launch::async, MoveToNode_Raw);
}
